src/content_stream.cpp

```diff
@@ -13,19 +13,29 @@
 // limitations under the License.
 
 #include "content_stream.h"
+#include "socket.h"
 
 #include "dap/io.h"
 
 #include <string.h>   // strlen
 #include <algorithm>  // std::min
+#include <memory>     // std::dynamic_pointer_cast
 
 namespace dap {
 
 ////////////////////////////////////////////////////////////////////////////////
 // ContentReader
 ////////////////////////////////////////////////////////////////////////////////
+static const char* HEADER_TERMINATOR = "\r\n\r\n";
+
 ContentReader::ContentReader(const std::shared_ptr<Reader>& reader)
-    : reader(reader) {}
+    : reader(reader) {
+  auto close_on_invalid_reader =
+      std::dynamic_pointer_cast<CloseOnInvalidReader>(reader);
```

Comment:
Good to know. I did start with that approach but I found that would require adding defaulted parameters in a few other functions that lead to the construction of `ContentReader`:
- [Session::connect](https://github.com/google/cppdap/blob/main/src/session.cpp#L72)
- [Session::bind](https://github.com/google/cppdap/blob/main/include/dap/session.h#L444)

This would allow a server to indicate it would like the close on invalid behavior [as in the example](https://github.com/google/cppdap/blob/main/examples/simple_net_client_server.cpp#L33)

Another option I considered is a new virtual method on `Reader`, but I believe that might break binary compatibility because `Reader` is in the public API, and adding a virtual method to an interface users may implement would break binary compatibility, right?

Shall I go ahead and add the default parameter to `Session::bind` and `Session::connect` then?

Reply:
You could add it as a property of the `Session`. Either as a defaulted constructor argument or a setter. Then there's only one place it needs to be passed / set.